### PR TITLE
feat: Save last validator timestsamps

### PR DESCRIPTION
**Pull request by deluca-mike**

### CODE REVIEW

File: src/MinterGateway.sol [hunk 1]
```diff
@@ -1078,58 +1081,89 @@ contract MinterGateway is IMinterGateway, ContinuousIndexing, ERC712Extended {
...
         bytes[] calldata signatures_
-    ) internal view returns (uint40 minTimestamp_) {
-        uint256 threshold_ = updateCollateralValidatorThreshold();
-
+    ) internal returns (uint40 minTimestamp_) {
         minTimestamp_ = uint40(block.timestamp);
 
+        uint256 validCount_;
+
         // Stop processing if there are no more signatures or `threshold_` is reached.
```
Comment (toninorair) [inline src/MinterGateway.sol:1089]: adjust comment - processing is not stopped anymore if threshold is reached

**Review by toninorair — COMMENTED** [review]
great work, very neat

File: src/MinterGateway.sol [hunk 2]
```diff
@@ -1066,58 +1074,88 @@ contract MinterGateway is IMinterGateway, ContinuousIndexing, ERC712Extended {
...
+                ++validCount_;
             }
         }
 
-        // NOTE: Due to STACK_TOO_DEEP issues, we need to refetch `requiredThreshold_` and compute the number of valid
-        //       signatures here, in order to emit the correct error message. However, the code will only reach this
-        //       point to inevitably revert, so the gas cost is not much of a concern.
-        if (threshold_ > 0) {
-            uint256 requiredThreshold_ = updateCollateralValidatorThreshold();
+        uint256 requiredThreshold_ = updateCollateralValidatorThreshold();
```
Comment (PierrickGT) [inline src/MinterGateway.sol:1109]: Can be inlined since in most cases we won't revert.

File: src/MinterGateway.sol [hunk 3]
```diff
@@ -1066,58 +1074,88 @@ contract MinterGateway is IMinterGateway, ContinuousIndexing, ERC712Extended {
...
 
-            unchecked {
-                // NOTE: By this point, it is already established that `threshold_` is less than `requiredThreshold_`.
-                revert NotEnoughValidSignatures(requiredThreshold_ - threshold_, requiredThreshold_);
-            }
-        }
+        if (validCount_ < requiredThreshold_) revert NotEnoughValidSignatures(validCount_, requiredThreshold_);
```
Comment (PierrickGT) [inline src/MinterGateway.sol:1111]:
```suggestion
        if (validCount_ < updateCollateralValidatorThreshold()) revert NotEnoughValidSignatures(validCount_, updateCollateralValidatorThreshold());
```

File: src/MinterGateway.sol [hunk 4]
```diff
@@ -1066,58 +1074,88 @@ contract MinterGateway is IMinterGateway, ContinuousIndexing, ERC712Extended {
...
+    /**
+     * @dev    Checks that a signature is a valid validator signature.
+     * @param  minter_       The address of the minter.
+     * @param  collateral_   The amount of collateral.
+     * @param  retrievalIds_ The list of outstanding collateral retrieval IDs to resolve.
+     * @param  metadataHash_ The hash of metadata of the collateral update, reserved for future informational use.
+     * @param  validator_    The address of a validator.
+     * @param  timestamp_    The timestamp for the collateral update signature.
+     * @param  signature_    The signature from the validator.
+     * @return valid_        Whether the signature is a valid validator signature.
```
Comment (PierrickGT) [inline src/MinterGateway.sol:1123]:
```suggestion
     * @return valid_        Whether the signature is a valid validator signature or not.
```

File: src/MinterGateway.sol [hunk 5]
```diff
@@ -1066,58 +1074,88 @@ contract MinterGateway is IMinterGateway, ContinuousIndexing, ERC712Extended {
...
+
+        // Check that ECDSA or ERC1271 signatures for given digest are valid.
+        valid_ = SignatureChecker.isValidSignature(
+            validator_,
+            _getUpdateCollateralDigest(minter_, collateral_, retrievalIds_, metadataHash_, timestamp_),
+            signature_
+        );
+
+        // Save the last signature timestamp for the minter and validator combination.
+        _lastSignatureTimestamp[minter_][validator_] = timestamp_;
```
Comment (PierrickGT) [inline src/MinterGateway.sol:1159]:
```suggestion
        if (!isValidatorApprovedByTTG(validator_)) {
            valid_ = false;
        } else {
            // Check that ECDSA or ERC1271 signatures for given digest are valid.
            valid_ = SignatureChecker.isValidSignature(
                validator_,
                _getUpdateCollateralDigest(minter_, collateral_, retrievalIds_, metadataHash_, timestamp_),
                signature_
            );

            // Save the last signature timestamp for the minter and validator combination.
            _lastSignatureTimestamp[minter_][validator_] = timestamp_;
        }
```

**Review by PierrickGT — APPROVED** [review]

File: src/MinterGateway.sol (hunk 2 above)
Comment (deluca-mike) [reply src/MinterGateway.sol:1109]: Gets multi-lined longer then, and I don't think it optimizes.

File: src/MinterGateway.sol (hunk 3 above)
Comment (deluca-mike) [reply src/MinterGateway.sol:1111]: Gets multi-lined longer then, and I don't think it optimizes.

File: src/MinterGateway.sol (hunk 5 above)
Comment (toninorair) [reply src/MinterGateway.sol:1159]: having one code block is easier for the reader of code vs if {} else {}, but it made me realize that this code can be unsafe

Comment (toninorair) [reply src/MinterGateway.sol:1159]: ` _lastSignatureTimestamp[minter_][validator_] = timestamp_;` should be done only for valid signatures 
@deluca-mike @PierrickGT 
more thought is needed here

Comment (PierrickGT) [reply src/MinterGateway.sol:1159]: > having one code block is easier for the reader of code vs if {} else {}, but it made me realize that this code can be unsafe

`valid_` being a named returned variable, it should be assigned and not implicitly returned.
Or if we want to return, we should also do it for the else case.

Comment (PierrickGT) [reply src/MinterGateway.sol:1159]:
```suggestion
        if (!isValidatorApprovedByTTG(validator_)) return false;

        // Check that ECDSA or ERC1271 signatures for given digest are valid.
        bool isValid_ = SignatureChecker.isValidSignature(
            validator_,
            _getUpdateCollateralDigest(minter_, collateral_, retrievalIds_, metadataHash_, timestamp_),
            signature_
        );

        if (isValid_) {
            // Save the last signature timestamp for the minter and validator combination.
            _lastSignatureTimestamp[minter_][validator_] = timestamp_;
        }

        return isValid_;
```

**Pull request merged**